File: 2456-construct-smallest-number-from-di-string/2456-construct-smallest-number-from-di-string.cpp
class Solution {
public:
    string smallestNumber(string pattern) {
        int count = 0;
        stack<int> st;
        string ans = "";
        for(int i = 0; i <= pattern.size(); i++){

            count++;
            if(pattern[i] == 'I' || i == pattern.size()){
                ans.push_back(count + '0');
                while(!st.empty()){
                    ans.push_back(st.top());
                    st.pop();
                }
            }
            else if(pattern[i] == 'D'){
                st.push(count + '0');
            }
        }
        return ans;
    }
};

File: 2456-construct-smallest-number-from-di-string/2456-construct-smallest-number-from-di-string.java
class Solution {
    public String smallestNumber(String pattern) {
        Deque<Integer> st = new ArrayDeque<>();
        StringBuilder ans = new StringBuilder();
        int cnt = 0;
        for(int i = 0; i <= pattern.length(); i++){
            cnt++;
            if(i == pattern.length() || pattern.charAt(i) == 'I'){
                ans.append(cnt);
                while(!st.isEmpty()){
                    ans.append(st.pop());
                }
            }
            else{
                st.push(cnt);
            }
        }
        return ans.toString();
    }
}